### ISSUE TITLE
RFC: Add POSIX shared untrusted memory support

### ISSUE THREAD

**Issue by llly** [gramineproject/gramine#757]
## Summary

As a sub-project of Communication with devices(#353), this RFC proposes a generic way to enable shared memory with host devices or processes from within the SGX (Gramine) enclave through POSIX shared memory.

Add new item to Manifest syntax: 
```
sys.insecure__shared_memory = "[none|passthrough]"
(Default: "none")
```
- none: default value, which means that shared untrusted memory is disabled, the same as current behavior in Gramine.
- passthrough: mount `/dev/shm` on host to `/dev/shm` in Gramine, which is folder contains shared memory file. Shared memory is created in untrusted memory, and accessible by its child, another Gramine or native process. Emulate shared memory of native app. the memory regions can be shared between the SGX enclave and an arbitrary host device.

Note that System V shared memory(`shmget`, `shmat`, etc.) is not in the scope.

## Scenarios
A designed scheme of using POSIX shared memory in LibOS to communicate with host devices is the following:
1. Application in Gramine creates  POSIX shared memory objects by Libc call `int shm_open(const char *name, int oflag, mode_t mode);`.  A nonnegative file descriptor is returned. Libc calls syscall `openat()`. Host create file `/dev/shm/{name}`.`/dev/shm` is a tmpfs provided by host, visible for application in Gramine and host.
2. Application in Gramine sets the size of memory-mapped file by syscall `int ftruncate(int fd, off_t length);`.
3. Application in Gramine map the file to the untrusted shared memory by `mmap(MAP_SHARED, fd)`.
4. Application in Gramine access untrusted shared memory.
5. Application in Gramine notify untrusted shared memory to host devices through a certain `ioctl` with the pointer in arg.
6. Host devices access untrusted shared memory.
7. Application in Gramine calls Libc call `int shm_unlink(const char *name);`. Libc calls syscall `unlink()`.  LibOS  remove the file `/dev/shm/{name}` and de-allocates untrusted shared memory.

A general scheme of using POSIX shared memory that server and client both `mmap("/dev/shm/{name}")` is also supported.

## Design
* In LibOS filesystem, introduce new filesystem type `shm_builtin_fs` .
    * `struct libos_fs shm_builtin_fs = { .name   = "shm", .fs_ops = &shm_fs_ops, .d_ops  = &shm_d_ops,};` 
    * `shm_fs_ops` supports limited operations `.mount`, `.mmap`, `.truncate`, `.hstat`.
    * `shm_d_ops` supports limited operations `.open`, `.lookup`, `.creat`, `.hstat`, `.unlink`.
    * The operations basically work as `chroot` operations do. `PAL_OPTION_PASSTHROUGH` option is used in Pal function called by `.open`, `.creat` and `.unlink`.

Main changes: Add `fs/shm/fs.c`, add `shm_builtin_fs` to `g_builtin_fs`. `mount_fs(libos_mount_params)` with `shm` type and `"/dev/shm"` of `.path` and `.uri` if enabled in manifest.

* Introduce two non-overlapping address ranges: private range and shared range
    * For LibOS: shared range is only used for `mmap(MAP_SHARED, fd of "/dev/shm/{name}")`, private range is used for `mmap(MAP_PRIVATE)` and other `mmap(MAP_SHARED)`. Shared range is also managed by existing VMA bookkeeping.
    * For common PAL: existing `g_pal_public_state.user_address` becomes private_range, plus a new `shared_range`. `void _PalGetAvailableUserAddressRange(void** out_start, void** out_end);` changed to `void _PalGetAvailableUserAddressRange(void** out_private_start, void** out_private_end, void** out_shared_start, void** out_shared_end);` to get the ranges.
    * For Linux PAL: private range is e.g. `0x0 - 1TB` and shared range is e.g. `1TB - 2TB`. (or other ranges)
    * For Linux-SGX PAL: private range is enclave memory and shared range is e.g. `1TB - 2TB` of untrusted memory (actually, a 1TB-sized region allocated with host's `mmap()` to reserve this memory, region location can be changed if enclave size is larger than 1TB)

Main changes in: `libos_mmap.c`, `pal_main.c`, `pal_file.c`. `libos_syscall_mmap()` checks `libos_fs` name. When name is `"shm"`, `shared_range` is used for VMA bookkeeping and mmap. Allowing `addr != NULL` in Linux-SGX PAL `file_map` functions to map host file to `shared_range`.

## Implementation
pseudo code for the structure and function changes:  
 * pal.h:
```
struct pal_public_state {
    ...
    void* private_user_address_start; /*!< Private user address range start */
    void* private_user_address_end;   /*!< Private user address range end */
    void* shared_user_address_start;  /*!< Shared user address range start */
    void* shared_user_address_end;    /*!< Shared user address range end */
    ...
}

```
* pal_internal.h:
```
#define PUBLIC_HEAP_MIN 0x10000000000ULL /* 1TB */
#define PUBLIC_HEAP_MAX 0x20000000000ULL /* 2TB */

void _PalGetAvailableUserAddressRange(void** out_private_start, void** out_private_end, void** out_shared_start, void** out_shared_end);
```
* LibOS:
```
struct libos_fs_ops shm_fs_ops = {
    .mount      = &shm_mount,
    .mmap       = &chroot_mmap,
    .hstat      = &generic_inode_hstat,
    .truncate   = &chroot_truncate,
    .poll       = &generic_inode_poll,
};

struct libos_d_ops shm_d_ops = {
    .open    = &chroot_open,
    .lookup  = &chroot_lookup,
    .creat   = &chroot_creat,
    .stat    = &generic_inode_stat,
    .readdir = &chroot_readdir,
    .unlink  = &chroot_unlink,
};

struct libos_fs shm_builtin_fs = {
    .name   = "shm",
    .fs_ops = &shm_fs_ops,
    .d_ops  = &shm_d_ops,
};

mount_sys()
      mount_fs(&(struct libos_mount_params){
        .type = "shm",
        .path = "/dev/shm/",
        .uri = "file:/dev/shm/",
    });

libos_syscall_mmap(addr, length, prot, flags, fd, offset)
    hdl = get_fd_handle(fd)
    if (flags & MAP_SHARED && hdl && !strcmp(hdl->fs->fs_ops->name(), "shm") )
        user_address_start/end = public_address_start/end
        alloc_type = PAL_ALLOC_SHARED
    else
        user_address_start/end = private_address_start/end
        alloc_type = 0
    /*following not changed*/
    bkeep_mmap(address_range, addr ...) /* to validate or choose addr */
    if (!hdl) /* unreachable in the shm file scenarios */
        PalVirtualMemoryAlloc(alloc_type)
    else
        ret = hdl->fs->fs_ops->mmap() /*chroot_mmap and generic_emulated_mmap */

chroot_mmap() /*not changed*/
    PalStreamMap()
        ops->map() /*only file_map*/
```
* Linux-SGX Pal:
```
void _PalGetAvailableUserAddressRange(void** out_private_start, void** out_private_end, void** out_shared_start, void** out_shared_end)
    *out_private_start/end = g_pal_linuxsgx_state.heap_min/max;
    void* public_heap_min = PUBLIC_HEAP_MIN;
    _PalVirtualMemoryAlloc(&public_heap_min, 1TB, PAL_ALLOC_SHARED, PROT_NONE)
    *out_shared_start/end = public_heap_min/public_heap_min + 1TB

_PalVirtualMemoryAlloc()
    if (alloc_type & PAL_ALLOC_SHARED)
        int flags = PAL_MEM_FLAGS_TO_LINUX(alloc_type, prot | PAL_PROT_WRITECOPY); /* MAP_PRIVATE | MAP_NORESERVE | MAP_UNINITIALIZED */
        int linux_prot = PAL_PROT_TO_LINUX(prot);
        flags |= MAP_ANONYMOUS | MAP_FIXED;
        ocall_mmap_untrusted(linux_prot, flags, /*fd=*/-1, /*offset=*/0);
    else
        get_enclave_pages()

file_map()
    if ((mem in PUBLIC_HEAP_MIN/MAX range) && allowed_file )
        ocall_mmap_untrusted(&mem, size, PAL_PROT_TO_LINUX(prot), MAP_SHARED MAP_FIXED, handle->file.fd, offset);
        return
    /*following not changed*/
    get_enclave_pages()
    if (trusted file)
        copy_and_verify_trusted_file
    else
        ocall_pread
```

* Linux Pal:
```
void _PalGetAvailableUserAddressRange(void** out_private_start, void** out_private_end, void** out_shared_start, void** out_shared_end)
    *out_private_start/end = MMAP_MIN_ADDR/PUBLIC_HEAP_MIN
    public_heap_min = PUBLIC_HEAP_MIN
    _PalVirtualMemoryAlloc(&addr, 1TB, PAL_ALLOC_SHARED, PROT_NONE)
    *out_shared_start/end = public_heap_min / public_heap_min+1TB

_PalVirtualMemoryAlloc() /* can work. not changed. */
```

**Comment by mkow** [comment]
One more issue: I don't like that this makes _all_ app's attempts to mmap shared memory go into untrusted memory, this makes it almost impossible to use securely (because you'll expose some other, random mmaps to the host). Can we somehow whitelist which mmaps are going outside the enclave?

**Comment by boryspoplawski** [comment]
For host devices we could have a list of them and only mmaps of devices from this list would be mapped in untrusted memory. Well, we probably need such list anyway, since we need to somehow tell Gramine which FS path map to which host device.

**Comment by llly** [comment]
@boryspoplawski 
In this issue I focus on POSIX shared memory API, instead of general shared memory.
I'm trying to enable a scheme that
> Application in LibOS notify untrusted shared memory to host devices through a certain ioctl

It's how the device I want works. So `mmap` usage here is not `mmap(...,  <general-device-fd>, ...)`, but `mmap(..., shared-memory-device-fd>, ...)`
'standalone' vs 'passthrough' means POSIX shared memory invisible vs visible for other application. No difference from security perspective. "disallowing access" in my origin decryption is inaccurate. no 
Do we want more general `mmap(...,  <general-device-fd>, ...)` to work? 

@mkow If only POSIX shared memory API supported, It's OK. If general-device-fd supported, we definitely need a whitelist like `sys.insecure__shared_memory = { device = [ "/dev/shm/shm1", "/dev/fd2", ], }`.

**Comment by dimakuv** [comment]
I feel like there is a bit of confusion. Let me summarize what flows we want to enable:

1. **Anonymous mmap** -- `mmap(MAP_ANONYMOUS, /*fd=*/-1)` and `mmap(MAP_ANONYMOUS | MAP_SHARED, /*fd=*/-1)`. Classic mmap, always allocates enclave memory (yes, even with `MAP_SHARED` -- or maybe in this case Gramine should just return error, doesn't matter for our discussion).
2. **Host-device-backed mmap** -- `mmap(MAP_SHARED, /*fd=*/device_fd)`. This is the "easy"/"classic" way to allocate shared untrusted memory that is backed by (shared with) the host device like a GPU. This is the case that @boryspoplawski mentions. This case should also be supported in Gramine, but it looks like this particular proposal from @llly doesn't cover this case (maybe because it's less controversial and easier to implement than the next one).
3. **POSIX-shared-memory-object-backed mmap** -- `mmap(MAP_SHARED, /*fd=*/dev_shm_fd)`. This mmap is preceeded by smth like `dev_shm_fd = open("/dev/shm/region1")` and followed by smth like `ioctl(device_fd, MYDEVICE_SET_SHARED_MEM_IOCTL, &mmapped_addr)`. This is the "complex" way to allocate shared untrusted memory that is backed by (shared with) the host device. This is the case that @boryspoplawski didn't know about/got confused about.

I should not one assumption about the applications that we'll make in this work:
- Normally, applications like TensorFlow and PyTorch simply use `mmap(MAP_ANONYMOUS | MAP_SHARED, /*fd=*/-1)` and then perform `ioctl(device_fd, MYDEVICE_SET_SHARED_MEM_IOCTL, &mmapped_addr)`. This is what we observed in all applications tested in the wild (I'd say 3 or 4 different runtimes).
- However, such an *anonymous mmap* is way too generic and commonly seen in normal, non-device-communication applications. So we don't want to have logic in Gramine that "as soon as we see `MAP_SHARED`, we allocate this memory as shared untrusted". Also, sometimes applications even do `mmap(MAP_ANONYMOUS, /*fd=*/-1)` and still bind this memory to the device via the `ioctl()` -- this is some evil programming but we've observed it in PyTorch iirc.
- So we will require applications to have a **very explicit** intent "yes, I really want to allocate a memory region that will be used as shared untrusted memory". That's what we opt for `mmap(MAP_SHARED, /*fd=*/dev_shm_fd)`. So we will require applications to be re-written to use this instead of *anonymous mmap*. (It's not a big problem in reality, all these runtimes need some modifications anyway to work with SGX enclaves.)

> What's the point of split `standalone` vs `passthrough`? From security perspective they are the same (untrusted memory) and, due to this, I see no point in disallowing access for other processes (including native).

@boryspoplawski I think the reasoning was to revoke these powers from the child SGX enclaves. Imagine a Python script that communicates with the device but also periodically spawns children to do some nonsense like `os.system("uname")`. It's better to prevent the child enclaves from accidentally touching this shared memory.

On the other hand, I also agree that we don't win much from this split on `standalone` vs `passthourgh`. Also, we complicate the Gramine logic and user experience. So I agree with Borys that this may not be a good idea -- a lot of complications for very incremental security.

> For host devices we could have a list of them and only mmaps of devices from this list would be mapped in untrusted memory. Well, we probably need such list anyway, since we need to somehow tell Gramine which FS path map to which host device.

@boryspoplawski @mkow You're talking about the case of *host-device-backed-mmap*. As @llly mentioned, this GitHub issue is not discussing this case. In reality, this case is already half-supported by Gramine. By adding something like this in the manifest:
```
fs.mounts = [
  { path = "/dev/mydevice", uri = "dev:/dev/mydevice" },  # note the `dev:` URI prefix
]
```

This syntax creates a device filename, so all host devices will need to be whitelisted using this syntax anyway. The only currently missing piece of the puzzle to work properly with such devices is the *host-device-backed-mmap*. (Recall that `open`/`read`/`write`/`close` already functions, and `ioctl` support is a separate PR: https://github.com/gramineproject/gramine/pull/671)

**Comment by dimakuv** [comment]
> Only one `passthrough` mode that shared memory is created in untrusted memory, but is invisible by another Gramine or native process or its child.

But why? I and Borys actually argued against "hiding" those regions of untrusted memory from other Gramine processes (e.g., children) -- Gramine should keep these regions in children as well.

> How can I update RFC? Edit the origin text?

Yes, you can just edit the original text.

**Comment by boryspoplawski** [comment]
@dimakuv Thanks for the detailed description, was super helpful!!

@llly Why exactly do you want to unshare this memory from other processes? Is it just for "sanity"/"because they don't need it" (more or less like @dimakuv described)?

**Comment by llly** [comment]
Got it . Only one passthrough mode that emulate POSIX shared memory of native app.

**Comment by llly** [comment]
Updated design, added Pal interface change and pseudo code.

**Comment by dimakuv** [comment]
@llly My feedback on the new design. Overall looks good to me.

> `sys.insecure__shared_memory = "[none|passthrough]"`

Now I think that we don't need this additional manifest option, since the user will have to mount the FS point `type="shm"` anyway explicitly. So it's just unnecessary duplication. The only usefulness of this feature is that it is easier to "notice" for manifest readers... @mkow @boryspoplawski @kailun-qin @guzongmin What do you think?

> the operations reuse `chroot` operations except `.mount`

Please don't *reuse* operations from chroot. This is confusing from the design point of view (shared memory uses "passthrough-files" logic? sounds weird). Also, I believe some quirks and corner cases of chroot operations will not be needed in the new "shm" FS.

So please use the chroot codebase as the basis for the new "shm", but keep the code separate (even if it will be almost 100% copy-paste).

> For common PAL: existing `g_pal_public_state.user_address` becomes private_range, plus a new `shared_range`. `void _PalGetAvailableUserAddressRange(void** out_start, void** out_end);` changed to `void _PalGetAvailableUserAddressRange(void** out_private_start, void** out_private_end, void** out_public_start, void** out_public_end);` to get the ranges.

I would prefer to make it more generic and to allow specifying a bunch of ranges. So smth like this:
```c
enum pal_memory_range_type {
    PAL_MEMORY_RANGE_PRIVATE,
    PAL_MEMORY_RANGE_SHARED,
}

struct pal_memory_range {
    enum pal_memory_range_type type;
    void* start;
    void* end;
}

g_pal_public_state.memory_ranges[0] = { .type = PAL_MEMORY_RANGE_PRIVATE, .start = ..., .end = ... }
g_pal_public_state.memory_ranges[1] = { .type = PAL_MEMORY_RANGE_SHARED, .start = ..., .end = ... }
```

This will give us the flexibility to have more than 2 ranges (e.g., when the SGX enclave is allocated at `[1TB, 2TB)`, then we can have the shared ranges at `[0TB, 1TB)`, `[2TB, 3TB)`). Even though the initial version can always use just two ranges, as you describe.

> Impact: `ASAN_SHADOW_START` starts at `3TB` instead of `1.5TB`

Why do we modify it? Just skip the ASan's memory range, and use e.g. `4TB` to `8TB` for shared memory. Let's not move ASan memory anywhere, but instead work around it.

> `_PalGetAvailableUserAddressRange(void** out_start, void** out_end)`

I propose smth like:
```c
_PalGetAvailableMemoryRanges(struct pal_memory_range* ranges, size_t* ranges_size);
```

The caller provides the `ranges` array and `&ranges_array_size`, and the function populates this array and updates `*ranges_size` with how many ranges there are actually. If the array size is not sufficient, then the function returns an error.

> `PAL_ALLOC_RESERVE`

You seem to be abusing this flag for shared memory region? I would suggest to introduce a new flag like `PAL_ALLOC_SHARED`, instead of relying on this already-existing flag.

**Comment by kailun-qin** [comment]
> > `sys.insecure__shared_memory = "[none|passthrough]"`
> 
> Now I think that we don't need this additional manifest option, since the user will have to mount the FS point `type="shm"` anyway explicitly. So it's just unnecessary duplication. The only usefulness of this feature is that it is easier to "notice" for manifest readers...

IMO, it's not worth adding an unnecessary/duplicated manifest just for easier "notice"...

**Comment by llly** [comment]
@dimakuv Thanks for the comments. It's clear enough for implementation now. I'll update this RFC while implement it in case I miss something.

**Comment by boryspoplawski** [comment]
> Add new mount type "shm":

Why do we need a new in-manifest mount? From what I understand in Gramine path is kind of frozen - it must be `/dev/shm/`. As for host path (`uri`) it's 1) host dependent 2) probably constant for any given PAL. Also I would expect only one such mount to be present (at most). Maybe it's pointless to have it and just have it always if `sys.insecure__shared_memory` is used?

> sgx.allowed_files = [ "file:/dev/shm" ]

Why is this needed? This should be a completely new filesystem, unrelated to `chroot`, `protected`, `allowed` etc. files.

> I would prefer to make it more generic and to allow specifying a bunch of ranges. So smth like this:

I'm not sure about this. This will complicate VMA management. Would it be so useful that it's worth the effort?

> Why do we modify it? Just skip the ASan's memory range, and use e.g. 4TB to 8TB for shared memory. Let's not move ASan memory anywhere, but instead work around it.

+1, let's leave it where it is

**Comment by llly** [comment]
> Why do we need a new in-manifest mount?

Can integrate shared memory into current filesystem flow without an amount of hackings, so does `sgx.allowed_files`, which is checked in Pal. I understand your concern @boryspoplawski. Normal users don't need this `shared memory`. the specific case needs this `shared memory` , but don't need to configure it. Currently user can only map read-only file to private memory. Here I want it to be a generic feature that can map a file with full access to shared memory though no requirement to map other files except `/dev/shm/{name}`. 
>I would prefer to make it more generic and to allow specifying a bunch of ranges. So smth like this:
>>I'm not sure about this. This will complicate VMA management. Would it be so useful that it's worth the effort?

I think it's more about scalability. Current VMA management only mange VMA inside enclave. In this change It also mange VMA for shared memory outside of enclave.

**Comment by dimakuv** [comment]
Replying to @boryspoplawski. Also, @llly please see if my replies make sense.

> Why do we need a new in-manifest mount? 

Well, I agree that it will always have `/dev/shm/`. Then I propose to have a simpler variant:
```toml
fs.mounts = [ {type = "shm"} ]          # implies /dev/shm/ in both Gramine and on host
```

>> `sgx.allowed_files = [ "file:/dev/shm" ]`

> Why is this needed?

Yep, I think we can skip this if we open `/dev/shm/` files using the `PAL_OPTION_PASSTHROUGH` flag. See:
- https://github.com/gramineproject/gramine/blob/90ea5a52609a872c992135f9311a47aeacd5f2b1/pal/include/pal/pal.h#L268
- https://github.com/gramineproject/gramine/blob/74e74a87a1127124cf89bd487a021dc1ceb8fa75/libos/src/fs/libos_fs_encrypted.c#L161

We already use it for `encrypted` files.

> I'm not sure about this. This will complicate VMA management. Would it be so useful that it's worth the effort?

Wasn't it you who proposed to have more than just two memory ranges? But I don't mind, if we think that two memory ranges is enough for the foreseable future, then we can hard-code just two ranges.

**Comment by boryspoplawski** [comment]
> But having said that, I know wonder "so what's the problem" :) We already do such mounting for special mounts like /dev/ and /proc/, see e.g. 

Exactly. This sounds like it's not needed (at least judging from this RFC), but I'm not sure what are next steps (after this is implemented).

> Wasn't it you who proposed to have more than just two memory ranges? But I don't mind, if we think that two memory ranges is enough for the foreseable future, then we can hard-code just two ranges.

No idea, maybe, but I don't remember (nothing in this issue about it either). I would stick to two for now just for the sake of simplicity - we can extend them if needed. Just need to make the code flexible enough so it will be possible to extend them. @llly 

> I think it's more about scalability. Current VMA management only mange VMA inside enclave. In this change It also mange VMA for shared memory outside of enclave.

@llly It wouldn't be a problem, we could stash all of them together, split between multiple trees or anything else.

**Comment by dimakuv** [comment]
I added `P0` priority simply because the PR is ready for review and should be merged for Gramine v1.5
